Adding duck type concept page (#3434)
knative/docs: docs/concepts/README.md

# Duck Typing

> Knative uses duck typing to keep various components loosely coupled, but
> [what is duck typing](https://en.wikipedia.org/wiki/Duck_typing)? It is
> reasoning about a resource's control plane shape and behaviors based on a common
> definition of that contact. If a resource has the same fields in the same schema

Reviewer: ```suggestion
Knative enables [loose coupling](https://en.wikipedia.org/wiki/Loose_coupling) of its components by using [duck typing](https://en.wikipedia.org/wiki/Duck_typing). Duck typing means that the compatibility of a resource for use in a Knative system is determined by certain properties being present that can be used to identify the resource control plane shape and behaviors. These properties are based on a set of common definitions for different types of resources, called duck types. If a resource has the same fields in the same schema
```

Reviewer: You don't have to use this but I suggested some stuff that made this make more sense to me 🙂

Author: the whole point I am drafting it and showing it :D make it make sense to others...

> locations as the contract specifies, and the control/data plane behaviors as the

Reviewer 2: ```suggestion
locations as the contract specifies, and the control/data plane behaves as the
```
Maybe behaves? I could be wrong.

> locations as the contract specifies, and the control/data plane behaviors as the

Reviewer: ```suggestion
locations as the common definition specifies, and the same control or data plane behaviors as the
```

> contract specifies, Knative can use that resource as if it is the generic duck

Reviewer: ```suggestion
common definition specifies, Knative can use that resource as if it is the generic duck
```

Reviewer: I'd suggest changing this just because the term contract seems very abstract when we're already explaining some pretty abstract stuff, but if contract here means something specific feel free to ignore me

> type with little specific knowledge about the resource type. Some resources may
> choose to opt-in or be opt'ed-in to multiple duck types.

Reviewer: ```suggestion
type, without specific knowledge about the resource type. Some resources may
choose to opt-in to multiple duck types.
```

Reviewer: I'd explain what it means to "opt in" better - does this just mean that they can use properties that would make them acceptable as more than one duck type?

Author: it means I can make a third party resource into a duck type (if it conforms to the shape) without the third party realizing this. the double hop integration

> <!-- TODO: point to Discovery ClusterDuckType documentation. -->

Reviewer: I'd make Discovery a different topic to Duck types under the Concepts folder, like I mentioned in the other comment, re not having this all in the index file. I think it'll be easier to read if it's a page per topic.

> The most basic usage of duck typing in Knative is our usage of object refs in
> resource specs to "point" to another resource. The contract of the object
> holding the ref will prescribe the expected duck type of the "pointee".

Reviewer: ```suggestion
A fundamental use of duck typing in Knative is the use of object references in
resource _specs_ to point to another resource. The definition of the object
containing the reference prescribes the expected duck type of the resource being referenced.
```

> For example, take this example Knative resource `pointer` that is pointing to a
> `Dog` named `pointee`:

Reviewer: ```suggestion
In the following example, a Knative `Example` resource named `pointer` references a
`Dog` resource named `pointee` in its spec:
```

```yaml
apiVersion: sample.knative.dev/v1
kind: Example
metadata:
  name: pointer
spec:
  size:
    apiVersion: extension.example.com/v1
    kind: Dog
    name: pointee
```

Assume the expected shape of a "Sizable" duck type is that in the status, we
expect the following schema shape:

```yaml
<standard metadata>
<spec ignored for Sizable>
status:
  height: <in inches>
  weight: <in pounds>

> ```

Reviewer: Since we don't talk about schemas anywhere else really until now, it might be good to explain what this actually is in terms of Knative resources? If the YAML example is basically an `Example` resource the same as the one above, this isn't obvious from the explanation, because it's not obvious that a schema and a resource are the same thing. I'd stick with just saying "resource" to avoid confusion if that's the case. I also think we should never abbreviate examples with things like `<standard metadata` because we're assuming a reader is more familiar with this than they might actually be, and again it's not clear if its the same thing as the resource metadata above.

Reviewer: Reading through this bit and the Human example bit I get what you're saying, but I think the way it's written is sort of confusing. I don't have time really to rewrite/edit it all for now, so I'm happy enough to merge as is when you're ready, but let's maybe put it on our to do list to revisit how we're explaining this later to see if it can be clearer 🙂

Author: Sure! I was just rolling with one concept as a start

Now the instance of `pointee` could look like this:

```yaml
apiVersion: extension.example.com/v1
kind: Dog
metadata:
  name: pointee
spec:
  owner: Smith Family
  etc: more here
status:
  lastFeeding: 2 hours ago
  hungry: true
  age: 2
  height: 27
  weight: 70
```

> When the `Example` resource needs to do it's work, it only acts on the

Reviewer 2: ```suggestion
When the `Example` resource needs to do its work, it only acts on the
```

information included in the "Sizable" duck type shape, and the `Dog`
implementation is free to have the information that makes the most sense for
that resource. The power of duck typing is apparent when we extend the system
with a new type, say, `Human`. Assuming the new resource adheres to the contract
set by the "Sizable".

```yaml
apiVersion: sample.knative.dev/v1
kind: Example
metadata:
  name: pointer
spec:
  size:
    apiVersion: people.example.com/v1
    kind: human
    name: pointee
---
apiVersion: people.example.com/v1
kind: Human
metadata:
  name: pointee
spec:
  etc: even more here
status:
  college: true
  hungry: true
  age: 22
  height: 62
  weight: 120
```

The `Example` resource was able to do the logic it is set to do without explicit
knowlage of `Human` or `Dog`.

## Knative Duck Types

> Knative defines Several duck type contracts that are in use across the project:

Reviewer 2: ```suggestion
Knative defines several duck type contracts that are in use across the project:
```
